BJS2-27199 - [Experienced] Покупка премиум доступа

dmitridorje opened the pull request:
Реализован функционал премиум-подписки (согласно описанию в задаче) и промоушн-подписки. В текущей реализации промоушн работает следующим образом. У пользователя есть возможность приобрести три варианта подписки: GENERAL (на 100 просмотров, приоритет 1), ADVANCED (на 500 просмотров, приоритет 2) и PREMIUM (на 750 просмотров, приоритет 3). Каждый вариант продвижения может быть доступен для поднятия либо профиля, либо событий, принадлежащих пользователю. При варианте PREMIUM также активируется таргетирование, при котором поднятие профиля или события будет происходить в том случае, если страна запрашивающего совпадает со страной того, кто приобрел промоушн. "Высота" подъёма также определяется приоритетом подписки - чем выше приоритет, тем выше профиль или событие будут в поиске.

dmitridorje committed:
first commit

dmitridorje committed:
Реализован функционал премиум и промоушн подписок

dmitridorje committed:
Добавлена возможность одновременной покупки двух вариантов промоушн-подписки

Rashit86 commented on src/main/java/school/faang/user_service/service/PremiumService.java:
```diff
@@ -0,0 +1,72 @@
+            throw new ExistingPurchaseException("User already has an active premium subscription");
+        }
+
+        payPremium(premiumPeriod);
```
Если платеж не прошел, то все равно идет сохранение премиума?

mikhailov26 replied:
+

dmitridorje replied:
Добавил соответствующие проверки, исключения и тесты.

Rashit86 commented on src/main/java/school/faang/user_service/service/PromotionService.java:
```diff
@@ -0,0 +1,73 @@
+            }
+        });
+
+        payPromotion(type);
```
promoition сохраняется даже если не прошел платеж?

dmitridorje replied:
Добавил соответствующие проверки, исключения и тесты.

Rashit86 commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,124 @@
+                .orElseGet(Collections::emptyList);
+    }
+
+    private List<Event> getPriorityFilteredEvents(List<Event> filteredEvents, User callingUser) {
```
Лучше отрефакторить, примерно так
private List<Event> getPriorityFilteredEvents(List<Event> filteredEvents, User callingUser) {
        return filteredEvents.stream()
                .sorted(Comparator
                        .comparing((Event event) -> calculateCountryPriority(event, callingUser))
                        .thenComparing(event -> calculatePriorityLevel(event)))
                .toList();
    }

    private int calculateCountryPriority(Event event, User callingUser) {
        User eventOwner = event.getOwner();
        if (eventOwner.getPromotions() == null || eventOwner.getPromotions().isEmpty()) {
            return 1;
        }

        Promotion targetPromotion = getTargetPromotion(event);
        if (targetPromotion == null || (targetPromotion.getPriorityLevel() == 3
                && !eventOwner.getCountry().equals(callingUser.getCountry()))) {
            return 1;
        }

        return 0;
    }

    private int calculatePriorityLevel(Event event) {
        User eventOwner = event.getOwner();
        if (eventOwner.getPromotions() == null || eventOwner.getPromotions().isEmpty()) {
            return 0;
        }

        Promotion targetPromotion = getTargetPromotion(event);
        return targetPromotion != null ? -targetPromotion.getPriorityLevel() : 0;
    }

dmitridorje replied:
Done!

Rashit86 commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -0,0 +1,123 @@
+                .orElseGet(Collections::emptyList);
+    }
+
+    private List<User> getPriorityFilteredUsers(List<User> filteredUsers, User callingUser) {
```
Тоже лучше отрефакторить

mikhailov26 replied:
+

dmitridorje replied:
Done!

Rashit86 commented on src/test/java/school/faang/user_service/service/event/EventServiceTest.java:
```diff
@@ -0,0 +1,140 @@
+    private EventMapper eventMapper = Mappers.getMapper(EventMapper.class);
+
+    @Mock
+    private List<EventFilter> eventFilters;
```
можно было мокнуть конкретные реализации фильтров, и из моков собрать List, как в видео Влада

mikhailov26 replied:
+

dmitridorje replied:
Done!

mikhailov26 commented:
В целом ок, задачка прям большая, очень хороший ПР получился, в данном случае согласен с комментариями Рашита, но прям особое внимание обратить на вариант, когда платёж не прошёл и в видео на платформе рекомендую ознакомиться с роликом с названием Покупка премиум доступа - дизайн (дополнительно)

dmitridorje committed:
добавлено исключение и соответствующие логика и тесты для неуспешных платежей;

выполнен рефакторинг сложных компараторов;

добавлены моки для фильтров в тестах

mikhailov26 commented:
Реши крнфликты и можешь мержить

dmitridorje committed:
Merge branch 'werewolf-master-stream6' of https://github.com/CorporationX/user_service into feature-BJS2-27199
